internal/blockchain/difficulty.go

```diff
@@ -212,20 +211,152 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+	// in order for the absolute calculations to exactly match the behavior of
+	// relative calculations.
+	//
+	// However, since the initial difficulty is reset with the agenda, no
+	// additional offsets are needed.
+	timeDelta := prevNode.timestamp - blake3Anchor.timestamp
+	heightDelta := prevNode.height - blake3Anchor.height
+
+	// Calculate the next target difficulty using the ASERT algorithm.
+	//
+	// Note that the difficulty of the anchor block is NOT used for the initial
+	// difficulty because the difficulty must be reset due to the change to
+	// blake3 for proof of work.  The initial difficulty comes from the chain
+	// parameters instead.
+	params := b.chainParams
+	nextDiff := standalone.CalcASERTDiff(params.WorkDiffV2Blake3StartBits,
+		params.PowLimit, int64(params.TargetTimePerBlock.Seconds()), timeDelta,
+		heightDelta, params.WorkDiffV2HalfLifeSecs)
+
+	// Prevent the difficulty from going higher than a maximum allowed
+	// difficulty on the test network.  This is to prevent runaway difficulty on
+	// testnet by ASICs and GPUs since it's not reasonable to require
+	// high-powered hardware to keep the test network running smoothly.
+	//
+	// Smaller numbers result in a higher difficulty, so imposing a maximum
+	// difficulty equates to limiting the minimum target value.
+	if b.minTestNetTarget != nil && nextDiff < b.minTestNetDiffBits {
+		nextDiff = b.minTestNetDiffBits
+	}
+
+	return nextDiff
+}
+
+// blake3WorkDiffAnchor returns the block to treat as the anchor block for the
+// purposes of determining how far ahead or behind the ideal schedule the
+// provided block is when calculating the blake3 target difficulty for the block
+// AFTER the passed previous block node.
+//
+// This function MUST only be called with the blake3 proof of work agenda active
+// and with the chain state lock held (for writes).
+func (b *BlockChain) blake3WorkDiffAnchor(prevNode *blockNode) *blockNode {
+	// Use the previously cached anchor when it exists and is actually an
+	// ancestor of the passed node.
+	//
+	// Note that it will always be set to the genesis block on networks where
+	// the agenda is always active.
+	anchor := b.cachedBlake3WorkDiffAnchor.Load()
+	if anchor != nil && anchor.IsAncestorOf(prevNode) {
+		return anchor
+	}
+
+	// Find the block just prior to the activation of the blake3 proof of work
+	// agenda from the perspective of the block AFTER the passed block.
+	//
+	// The state of an agenda can only change at a rule change activation
+	// boundary, so determine the final block of the rule change activation
+	// interval just prior to the interval the block AFTER the provided one
+	// falls in and then loop backwards one interval at a time until the agenda
+	// is no longer active.
+	rcai := int64(b.chainParams.RuleChangeActivationInterval)
+	svh := b.chainParams.StakeValidationHeight
+	finalNodeHeight := calcWantHeight(svh, rcai, prevNode.height+1)
+	candidate := prevNode.Ancestor(finalNodeHeight)
```

Comment:
This is theoretically true (and part of why the following checks help protect against misuse nil checks), but, in practice, due to the precondition that the function is required to be called with the agenda active, it can't ever be the genesis block or `nil`.

This is because either:
(a) the anchor was the genesis block already and thus the function would've returned it above; or
(b) the agenda must have gone through activation which required an interval for it to be locked in meaning the floor here would be the final node of the locked in interval.